yourbench/utils/inference_engine.py

```diff
@@ -1,122 +1,301 @@
-
-#     for i in tqdm(
-#         range(0, len(user_prompts), batch_size),
-#         total=total_batches,
-#         desc="Processing batches",
-#     ):
-#         batch_prompts = user_prompts[i : i + batch_size]
-#         batch_completions = litellm.batch_completion(
-#             model=f"{model_type}/{model_name}",
-#             messages=batch_prompts,
-#             **kwargs,
-#         )
-#         completions.extend(batch_completions)
-#     final_responses = []
-#     for response in completions:
-#         try:
-#             final_responses.append(response.choices[0].message.content)
-#         except Exception as e:
-#             logger.error(f"Error processing response: {e}")
-#             final_responses.append("")
-#     return final_responses
-
-
-async def _process_single_prompt(
-    session: aiohttp.ClientSession,
-    prompt: dict,
-    model_name: str,
-    api_base_url: str = None,
-    api_key: str = None,
-    semaphore: asyncio.Semaphore = None,
-    retry_attempts: int = 6,
-    timeout_seconds: int = 600,
-) -> dict:
-    async with semaphore:
-        for attempt in range(retry_attempts):
+GLOBAL_TIMEOUT = 3600
+
+# TODO: why do we need these lines
+# Optional: Customize success/failure callbacks
+litellm.success_callback = ["langfuse"]
+litellm.failure_callback = ["langfuse"]
+
+
+class EventLoopAwareCache(litellm.InMemoryCache):
+    """
+    A cache that maintains cache isolation between event loops.
+
+    LiteLLM caches LLM clients for better performance, but shares this cache across
+    event loops. LiteLLM uses httpx AsyncClient under the hood, and you should not
+    share AsyncClients across event loops.
+    - Same issue posted to LiteLLM: https://github.com/BerriAI/litellm/issues/7667
+    - Reference to httpx behavior: https://github.com/encode/httpx/issues/2473
+
+    The cache needs to manage a uuid identifier that won't be reused after an event
+    loop is terminated.
+    """
+
+    def get_cache(self, key, **kwargs):
+        new_key = self._get_event_loop_aware_cache_key(key)
+        return super().get_cache(new_key)
+
+    def set_cache(self, key, value, **kwargs):
+        new_key = self._get_event_loop_aware_cache_key(key)
+        return super().set_cache(new_key, value, **kwargs)
+
+    def _get_event_loop_aware_cache_key(self, key):
+        loop = asyncio.get_running_loop()
+        if not hasattr(loop, "_custom_identifier"):
+            loop._custom_identifier = uuid.uuid4()
+        loop_id = loop._custom_identifier
+        return f"{key}-{loop_id}"
+
+
+litellm.in_memory_llm_clients_cache = EventLoopAwareCache()
+
+
+@dataclass
+class Model:
+    model_name: str
+    request_style: str = "openai"
+    base_url: str = None
+    api_key: str = None
+    max_concurrent_requests: int = 8
+    inference_backend: str = "litellm"
+    provider: str = None
+    api_version: str = None
+
+
+@dataclass
+class InferenceCall:
```

Comment:
add docstring (what are the tags?)

Reply:
notably important if not all options are supported by all backends

Reply:
added explanation regarding tags. they're mostly for internal logging and cost tracking with langfuse